clients/include/testing_tpsv.hpp

```diff
@@ -0,0 +1,266 @@
+    char        char_uplo   = arg.uplo;
+    char        char_transA = arg.transA;
+    char        char_diag   = arg.diag;
+
+    rocblas_fill      uplo   = char2rocblas_fill(char_uplo);
+    rocblas_operation transA = char2rocblas_operation(char_transA);
+    rocblas_diagonal  diag   = char2rocblas_diagonal(char_diag);
+
+    rocblas_status       status;
+    rocblas_local_handle handle;
+
+    // check here to prevent undefined memory allocation error
+    if(N < 0 || !incx)
+    {
+        CHECK_ROCBLAS_ERROR(rocblas_set_pointer_mode(handle, rocblas_pointer_mode_host));
+        EXPECT_ROCBLAS_STATUS(
+            rocblas_tpsv<T>(handle, uplo, transA, diag, N, nullptr, nullptr, incx),
+            rocblas_status_invalid_size);
+        return;
+    }
+
+    size_t size_A   = size_t(N) * N;
+    size_t size_AP  = size_t(N) * (N + 1) / 2.0;
+    size_t abs_incx = size_t(incx >= 0 ? incx : -incx);
+    size_t size_x   = N * abs_incx;
+
+    // Naming: dK is in GPU (device) memory. hK is in CPU (host) memory
+    host_vector<T> hA(size_A);
+    host_vector<T> AAT(size_A);
+    host_vector<T> hAP(size_AP);
+    host_vector<T> hb(size_x);
+    host_vector<T> hx(size_x);
+    host_vector<T> hx_or_b_1(size_x);
+    host_vector<T> hx_or_b_2(size_x);
+    host_vector<T> cpu_x_or_b(size_x);
+    host_vector<T> my_cpu_x_or_b(size_x);
+
+    double gpu_time_used, cpu_time_used;
+    double rocblas_gflops, cblas_gflops, rocblas_bandwidth;
+    double rocblas_error;
+    double error_eps_multiplier    = 40.0;
+    double residual_eps_multiplier = 20.0;
+    double eps                     = std::numeric_limits<real_t<T>>::epsilon();
+
+    // allocate memory on device
+    device_vector<T> dAP(size_AP);
+    device_vector<T> dx_or_b(size_x);
+    CHECK_DEVICE_ALLOCATION(dAP.memcheck());
+    CHECK_DEVICE_ALLOCATION(dx_or_b.memcheck());
+
+    rocblas_init<T>(hA, true);
+
+    //  calculate AAT = hA * hA ^ T
+    cblas_gemm<T, T>(rocblas_operation_none,
+                     rocblas_operation_conjugate_transpose,
+                     N,
+                     N,
+                     N,
+                     T(1.0),
+                     hA,
+                     N,
+                     hA,
+                     N,
+                     T(0.0),
+                     AAT,
+                     N);
+
+    //  copy AAT into hA, make hA strictly diagonal dominant, and therefore SPD
+    for(int i = 0; i < N; i++)
+    {
+        T t = 0.0;
+        for(int j = 0; j < N; j++)
+        {
+            hA[i + j * N] = AAT[i + j * N];
+            t += rocblas_abs(AAT[i + j * N]);
+        }
+        hA[i + i * N] = t;
+    }
+    //  calculate Cholesky factorization of SPD matrix hA
+    cblas_potrf<T>(char_uplo, N, hA, N);
+
+    //  make hA unit diagonal if diag == rocblas_diagonal_unit
+    if(char_diag == 'U' || char_diag == 'u')
+    {
+        if('L' == char_uplo || 'l' == char_uplo)
+            for(int i = 0; i < N; i++)
```

Comment:
I think you can use openmp for loops here if you want. Not sure if that is done in rcoBLAS or not.

Reply:
Didn't see any other openmp loops in rocBLAS so left it for now, it's just testing code anyway.

Reply:
We could parallelize CPU code with OpenMP to improve performance a little bit, which might become necessary at some point if the testing time gets too large.